File: c++/SerializeAndDeserializeTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        // preorder print tree
        string result = "";
        result = serializeHelper(root);
        return result;
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        int pos = 0;
        return deserializeHelper(data, pos);
        
    }

private:
    string serializeHelper(TreeNode* root){
        string result = "";
        if (root == nullptr){
            result = "#,";
        }
        else{
            result = to_string(root->val) + ",";
            result += serializeHelper(root->left);
            result += serializeHelper(root->right);
        }  
        return result;
        
    }
    
    TreeNode* deserializeHelper(string& data,int& pos){
        if (pos < data.size() && data[pos] == '#'){
            pos += 2;
            return nullptr;
        }
        else{
            int sep_pos = data.find(',', pos);
            int val = stoi(data.substr(pos, sep_pos - pos));
            TreeNode* root = new TreeNode(val);
            pos += sep_pos - pos + 1;
            root->left = deserializeHelper(data, pos);
            root->right = deserializeHelper(data, pos);
            return root;
        }
        
    }

};

// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.deserialize(codec.serialize(root));

File: python/serialize_and_deserialize_tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.

        :type root: TreeNode
        :rtype: str
        """
        result = ''
        if root is None:
            result += '#,'
            return result
        result += str(root.val) + ','
        result += self.serialize(root.left)
        result += self.serialize(root.right)
        return result

    def deserialize(self, data):
        """Decodes your encoded data to tree.

        :type data: str
        :rtype: TreeNode
        """
        sp = data[:-1].split(',')
        root, _ = self.deserializeHelper(sp, 0)
        return root

    def deserializeHelper(self, data, pos):
        if pos >= len(data) or data[pos] == '#':
            return None, pos + 1

        root = TreeNode(int(data[pos]))
        root.left, pos = self.deserializeHelper(data, pos + 1)
        root.right, pos = self.deserializeHelper(data, pos)

        return root, pos


# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))
